scripts/generate_rust_analyzer.py

```diff
@@ -0,0 +1,175 @@
+#!/usr/bin/env python3
+import sys
+import os
+from os import path
+import json
+
+if (len(sys.argv) != 5):
+    print("Not enough arguments")
+    sys.exit(-1)
+
+srctree = sys.argv[1]
+objtree = sys.argv[2]
+lib_src = sys.argv[3]
+bindgen_file = sys.argv[4]
+
+crate_list = []
+
+def find(list, name):
+    pos = -1
+    for x in list:
+        pos += 1
+        if x['display_name'] == name:
+            return pos
+
+def generate_deps(source, deps):
+    ret = []
+    for x in deps:
+        ret.append({"crate": find(source, x), "name": x})
+    return ret
+
+def generate_cfgs():
+    ret = []
+    with open( objtree + "/include/generated/rustc_cfg") as f:
+        for line in f:
+            line = line.replace('--cfg=', '')
+            line = line.replace('\n', '')
+            ret.append(line)
+    return ret
+
+def get_default_deps(source = crate_list):
+    return generate_deps(source, ["core", "alloc", "kernel"])
+
+def generate_crate(name, module, member = True, deps = get_default_deps(), cfg = generate_cfgs(), extra = {}, edition = "2018"):
```

Comment:
Careful with mutable default parameters (and function calls). In general, in Python, only use immutable default parameters (literals, typically) -- anything else should be `None` and handled in the implementation.

The reason is that default parameters are evaluated when the function definition is encountered, not when called, e.g. this:

```python
def f(l = []):
    l.append(42)
    print(l)

f()
f()
```

Prints:
```python
[42]
[42, 42]
```

(Yeah, Python got this wrong, it is a major pitfall)

Same for `get_default_deps()` etc. which take a global mutable default parameter.

Reply:
removed that with the classy way